python/pdo/contract/response.py

```diff
@@ -137,6 +362,86 @@ def __init__(self, request, response) :
             self.new_state_object.pull_state_from_eservice(self.enclave_service)
+            
+            # compute ids of blocks in the change set (used for replication)
+            self.new_state_object.compute_ids_of_newblocks(request.contract_state.component_block_ids)
+            self.replicator = Replicator(self.replication_params, self.storage_clients_for_replication, \
+                self.contract_id, self.new_state_object.changed_block_ids, request.enclave_service)
+
+    # -------------------------------------------------------
+    
+    @property
+    def commit_id(self):
+        if self.status and self.state_changed:
+            return (self.contract_id, self.new_state_hash, self.request_number)
+        else:
+            return None
+        
+    # -------------------------------------------------------
+    def replicate_change_set(self):
+        """replicate change set for the current contract state to all provisioned enclaves"""
+            
+        try:
+            self.replicator.replicate_change_set()
+        except ReplicationException as e:
+            logger.exception("Failed to replicate change set %s", str(e))
+            raise Exception("Failed to replicate change set %s", str(e))
+
+    # -------------------------------------------------------    
+    def commit_asynchronously(self, ledger_config, wait, dependency_list_txnids=[], dependency_list_commit_ids=[], use_ledger=True):
+        """Commit includes two steps: First, replicate the change set to all provisioned encalves. Second, 
+        commit the transaction to the ledger. In this method, we add a job to the replication queue to enable the first step. The job will 
+        be picked by a replicator thead which after replication will add a sumit txn job to the pending transactions queue for the 
+        second step """
+
+        #sanity check: ensure that commit_id is not None
+        if self.commit_id is None:
+            logger.error('Commiting a response that should not be committed. Perhaps the state did not change? Or a failed update?')
+            raise Exception('Commiting a response that should not be committed. Perhaps the state did not change? Or a failed update?')
+
+        #start threads for commiting response if not done before
+        if ContractResponse.__start_threads_for_commit__:
+            # start replicator threads
+            for i in range(Replicator.get_max_num_replicator_threads()):
+                Replicator.replication_executor.submit(Replicator.replication_worker)
+                
+            # start txn submisstion thread
+            if use_ledger:
+                ContractResponse.transaction_executor.submit(ContractResponse.transaction_submission_worker)
+                
+            ContractResponse.__start_threads_for_commit__ = False
+        
+        txn_params = dict()
+        txn_params['use_ledger'] = use_ledger
+        txn_params['ledger_config'] = ledger_config
```

Comment:
is use_ledger is false then why are we doing anything?
and shouldn't that situation be managed at a higher level?

Reply:
yes thank you, i have fixed this